### PR TITLE
Add constraints, TypeSchema, and TypeSchemaSubstitutions

### PR DESCRIPTION
**Pull request by aaronstanek**

### CODE REVIEW

File: rust/type_checker/src/type_schema_substitutions.rs [hunk 1]
```diff
@@ -0,0 +1,423 @@
+use crate::{
+    constraints::{Constraint, HasFieldConstraint, HasMethodConstraint, SubTagConstraint},
+    type_schema::TypeSchema,
+    GenericTypeId,
+};
+use std::collections::HashMap;
+
+pub struct TypeSchemaSubstitutions {
+    disjoint_set: HashMap<GenericTypeId, Option<GenericTypeId>>,
```
Comment (Nick-Mazuk) [inline rust/type_checker/src/type_schema_substitutions.rs]: What is the reason for making the value an option?

By default, a `HaspMap` will return an option. So unless there's a specific reason to have `disjoint_set.get(5)` return a `Some(None)`, we can just remove the option.

File: rust/type_checker/src/type_schema_substitutions.rs [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+impl TypeSchemaSubstitutions {
+    pub fn new() -> Self {
+        Self {
+            disjoint_set: HashMap::new(),
+        }
+    }
+    pub fn insert_new_id(&mut self, typ: GenericTypeId) {
+        self.disjoint_set.insert(typ, None);
+    }
+    fn get_canonical_id_impl(&mut self, typ: GenericTypeId) -> Option<GenericTypeId> {
```
Comment (Nick-Mazuk) [inline rust/type_checker/src/type_schema_substitutions.rs]: It seems like each type constraint will get an ID. However, the same constraint may be given multiple IDs if the constraint was generated in multiple places. So here, you're trying to get the canonical ID for each constraint. That way you can later compare two constraints by simply comparing a number?

Then, are we assigning these IDs consecutively? For instance, the first constraint we find will have an ID of 0, the next 1, and so on?

If so, consider using a vector here. It'll be faster and the code will be simpler. And it's more intuitive to check if the ID exists—is it smaller than the length of the vector?

```rust
pub struct TypeSchemaSubstitutions {
    ids: Vec<GenericTypeId>,
}
```

If you do this, then `insert_new_id` would probably instead be the following:

```rust
pub fn get_new_id(&mut self) -> GenericTypeId {
    let id = self.ids.len();
    self.ids.push(id);
    id
}
```

Then by default, `self.ids[id]` will always return the canonical ID for `id`. No need to check if it has a parent or not. If you need to merge two IDs, then you can just overwrite one of them (which you're doing anyways).

The disadvantage of this approach is you'll essentially have a single object that can assign IDs, which means you cannot multithread that process. But I think multithreading is out of the scope for both the MVP and bootstrapped parser (JS can't really multithread, anyways).

File: rust/type_checker/src/type_schema_substitutions.rs [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+                method_name: "world".to_owned(),
+                method_type: type_c,
+            })],
+        );
+        schema
+            .constraints
+            .insert(type_c, vec![Constraint::HasReturnType(type_a)]);
+        substitutions.set_types_equal(type_a, type_b);
+        let new_schema = substitutions.apply_to_type_schema(schema);
+        for (_, constraint_vec) in new_schema.constraints {
```
Comment (Nick-Mazuk) [inline rust/type_checker/src/type_schema_substitutions.rs:298]: No logic in tests. That includes for loops.

Just write out each of the asserts for the values you want to test individually.

File: rust/type_checker/src/type_schema.rs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    pub fn new() -> Self {
+        Self {
+            next_id: 0,
+            constraints: HashMap::new(),
+            imports: HashMap::new(),
+        }
+    }
+    pub fn make_id(&mut self) -> GenericTypeId {
+        let return_value = self.next_id;
+        self.next_id += 1;
```
Comment (Nick-Mazuk) [inline rust/type_checker/src/type_schema.rs:22]: Correct me if I'm wrong, but it seems like if you're doing it this way, then we may end up with duplicate constraint IDs across the program. Wouldn't that provide the opportunity to have two different constraints get the same ID and create many bugs? If so, that's also a security concern because someone could exploit it.

**Review by Nick-Mazuk — CHANGES_REQUESTED** [review]

File: rust/type_checker/src/type_schema_substitutions.rs (hunk 1 above)
Comment (aaronstanek) [reply rust/type_checker/src/type_schema_substitutions.rs]: If a key maps to None, it means that the key is its own canonical id. If a key maps to Some, then the payload of Some is a better guess for the canonical id than the key. I'll add a comment explaining this.

File: rust/type_checker/src/type_schema_substitutions.rs (hunk 2 above)
Comment (aaronstanek) [reply rust/type_checker/src/type_schema_substitutions.rs]: I'll add some comments explaining all of this.

The ids refer to types, not constraints.

In the current implementation, type ids are be inserted consecutively, in increasing order, starting from zero. In this case, a vector would be substantially more performant than a HashMap. But if the implementation were to ever change in the future, then a vector might create issues.

Thinking about this, it is probably safe to use a vector here, and the performance benefits would be substantial. I'm going to make that change and add a test requiring that ids be inserted consecutively, in increasing order, starting from zero.

File: rust/type_checker/src/type_schema_substitutions.rs (hunk 3 above)
Comment (aaronstanek) [reply rust/type_checker/src/type_schema_substitutions.rs:298]: The issue that I ran into is that I wanted to assert that the constraint vectors of `type_a` and the `type_b` are concatenated and placed under a single key. Whether the final key is `type_a` or `type_b` is an implementation detail, and should not appear in the tests.

If you have a cool trick for doing this without logic or arithmetic that would be really handy.

File: rust/type_checker/src/type_schema.rs (hunk 4 above)
Comment (aaronstanek) [reply rust/type_checker/src/type_schema.rs:22]: The ids refer to types, not constraints.

The ids are scoped to the level of a function definition, record definitions, or tag union.

I'll add some comments explaining this.

File: rust/type_checker/src/type_schema_substitutions.rs (hunk 1 above)
Comment (Nick-Mazuk) [reply rust/type_checker/src/type_schema_substitutions.rs]: What if the value is just always the canonical ID? If a key maps to itself, then you know it's its own canonical ID. If it maps to something else, then you know that is it's canonical ID.

This makes looking up a canonical ID much simpler. Just lookup the value associated with that key.

File: rust/type_checker/src/type_schema_substitutions.rs (hunk 2 above)
Comment (Nick-Mazuk) [reply rust/type_checker/src/type_schema_substitutions.rs]: In general, as long as the API doesn't change, it gives you the flexibility to change the underlying data structures in the future without much issue.

File: rust/type_checker/src/type_schema_substitutions.rs (hunk 1 above)
Comment (aaronstanek) [reply rust/type_checker/src/type_schema_substitutions.rs]: The current implementation has amortized `O(1)`* time lookup and updates. The changes you propose would guarantee a non-amortized `O(1)` lookup time, but would slow updates to `O(n)`, where `n` is the number of types in the schema.

I think that it is important to clarify that when type `a` is set equal to type `b`, all other types equal to `a` must then be made equal to all other types equal to `b`. The [disjoint set data structure](https://en.wikipedia.org/wiki/Disjoint-set_data_structure) is designed for exactly this situation.

*Disjoint sets aren't exactly `O(1)` time. The data structure has a rather complicated time complexity, but from an engineering standpoint the distinction between `O(1)` and very nearly `O(1)` is meaningless.

Comment (Nick-Mazuk) [reply rust/type_checker/src/type_schema_substitutions.rs]: You can still use disjoint sets the way I'm proposing. It's actually the main way I've implemented them in the past. I've actually only heard them by the name "union find" before. It's also one of my favorite algorithms.

The main difference is that you'll have a simpler data structure, the code will be simpler, and it'll take up less memory.

```rust
#[derive(Default)]
pub struct TypeSchemaSubstitutions {
    ids: Vec<GenericTypeId>,
}

impl TypeSchemaSubstitutions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn get_new_id(&mut self) -> GenericTypeId {
        let id = self.ids.len();
        self.ids.push(id);
        id
    }

    pub fn get_canonical_id(&mut self, typ: GenericTypeId) -> GenericTypeId {
        let parent = self.ids[typ];
        if parent == typ {
            return typ;
        }
        self.ids[typ] = self.ids[parent];
        self.ids[typ]
    }
}
```

Because we guarantee inside this data structure that the IDs are consecutive and always within the bounds of the vector, we don't need to have all the out of bounds checks. You can still do that if you want by using `self.ids.get(typ)` instead.

Comment (aaronstanek) [reply rust/type_checker/src/type_schema_substitutions.rs]: I was getting tripped up by the terminology. In a union find / disjoint set data structure, each entry is mapped to a parent element, not a canonical element (Wikipedia uses "root" instead of "canonical"). But I do understand now what you mean by removing the option. I included to be more Rust-idiomatic, but on reconsidering, I'm not sure that it makes much of a difference. I'll remove it.

File: rust/type_checker/src/type_schema_substitutions.rs (hunk 3 above)
Comment (Nick-Mazuk) [reply rust/type_checker/src/type_schema_substitutions.rs:298]: Hmm…

I've thought of several ways to do this, but all of those would make the code more confusing.

If you needed to verify _every single constraint_, this loop might be the cleanest option unfortunately.

However, if you did not need to very every single constraint, I'd just pick the first elements from every item.

```rust
if let Constraint::HasField(field_constraint) = new_schema.constraints[0].1[0] {
    assert_eq!(
        substitutions.get_canonical_id(field_constraint.field_type),
        substitutions.get_canonical_id(type_a)
    );
}
```

One other thing, you've also already set `type_a` to equal `type_b` above.

```rust
substitutions.set_types_equal(type_a, type_b);
```

So checking the constraints equal both types is redundant. You can trust the `set_types_equal` method because you already wrote tests for that.

**Review by Nick-Mazuk — APPROVED** [review]
Approving since I think we're aligned on the fixes. That way you're unblocked from merging once it's ready.
